ts/parsers/paragraphparser.ts

```diff
@@ -1,37 +1,155 @@
-import { getValueAtPath } from "../helpers";
+import { checkPath, getValueAtPath } from "../helpers";
 import { ColorParser } from "./";
 
 import {
     PowerpointElement,
     TextAlignment,
     FontAttributes,
     Paragraph,
-    Content
+    Content,
+    List,
+    ListType
 } from "airppt-models-plus/pptelement";
 
 /**
  * Parse the paragraph elements
  */
 export default class ParagraphParser {
+    public static isList(paragraph): boolean {
+        return (
+            checkPath(paragraph, '["a:pPr"][0]["a:buAutoNum"]') ||
+            checkPath(paragraph, '["a:pPr"][0]["a:buChar"]')
+        );
+    }
+
+    public static getParagraph(paragraph): Paragraph {
+        const textElements = paragraph["a:r"] || [];
+        const content = textElements.map((txtElement) => {
+            return {
+                text: txtElement["a:t"] || "",
+                textCharacterProperties: this.determineTextProperties(
+                    getValueAtPath(txtElement, '["a:rPr"][0]')
+                )
+            };
+        });
+
+        return {
+            content: content,
+            paragraphProperties: this.determineParagraphProperties(paragraph)
+        };
+    }
+
+    public static getListlevel(paragraph): number {
+        const level = getValueAtPath(paragraph, '["a:pPr"][0]["$"]["lvl"]');
+
+        return level ? parseInt(level) : 0;
+    }
+
+    public static getListType(paragraph): ListType {
+        if (checkPath(paragraph, '["a:pPr"][0]["a:buAutoNum"]')) {
+            return ListType.Ordered;
+        }
+
+        return ListType.UnOrdered;
+    }
+
+    public static restructureList(list: List): List {
+        for (let i = 0; i < list.listItems.length - 1; i++) {
+            if (list.listItems[i + 1].list) {
+                list.listItems[i]["list"] = this.restructureList(list.listItems[i + 1].list);
+                list.listItems.splice(i + 1, 1);
+            }
+        }
+        return list;
+    }
+
     public static extractParagraphElements(paragraphs: any[]): PowerpointElement["paragraph"] {
         if (!paragraphs || paragraphs.length === 0) {
             return null;
         }
 
-        return paragraphs.map((paragraph) => {
-            const textElements = paragraph["a:r"] || [];
-            const content = textElements.map((txtElement) => {
-                return {
-                    text: txtElement["a:t"] || "",
-                    textCharacterProperties: this.determineTextProperties(getValueAtPath(txtElement, '["a:rPr"][0]'))
-                };
-            });
+        const paras = [];
+        const stack = [];
```

Comment:
```suggestion
        const parentNodes = [];
```

just a suggestion, feel free to ignore

Reply:
stack gives better code readability in a sense that you would know that this array is being used as a stack, since i am using `pop` function as well. so `stack.pop` looks better and clearer.